crates/key-value-azure/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use azure_data_cosmos::{
+    prelude::{AuthorizationToken, CollectionClient, CosmosClient, Query},
+    CosmosEntity,
+};
+use futures::StreamExt;
+use serde::{Deserialize, Serialize};
+use spin_core::async_trait;
+use spin_key_value::{log_error, Error, Store, StoreManager};
+use tokio::{
+    sync::{Mutex, OnceCell},
+    task,
+};
+
+pub struct KeyValueAzureCosmos {
+    key: String,
+    account: String,
+    database: String,
+    collection: String,
+    pub client: OnceCell<Arc<Mutex<CollectionClient>>>,
+}
+
+impl KeyValueAzureCosmos {
+    pub fn new(key: String, account: String, database: String, collection: String) -> Result<Self> {
+        Ok(Self {
+            key,
+            account,
+            database,
+            collection,
+            client: OnceCell::new(),
+        })
+    }
+}
+
+#[async_trait]
+impl StoreManager for KeyValueAzureCosmos {
+    async fn get(&self, name: &str) -> Result<Arc<dyn Store>, Error> {
+        let client = task::block_in_place(|| {
+            self.client.get_or_try_init(|| async {
+                let token =
+                    AuthorizationToken::primary_from_base64(&self.key).map_err(log_error)?;
+
+                let cosmos_client = CosmosClient::new(self.account.clone(), token);
+                let database_client = cosmos_client.database_client(self.database.clone());
+
+                let collection_client = database_client.collection_client(self.collection.clone());
+                Ok(Arc::new(Mutex::new(collection_client)))
+            })
+        })
+        .await?;
+
+        Ok(Arc::new(AzureCosmosStore {
+            _name: name.to_owned(),
+            client: client.clone(),
+        }))
+    }
+}
+
+struct AzureCosmosStore {
+    _name: String,
+    client: Arc<Mutex<CollectionClient>>,
+}
+
+#[async_trait]
+impl Store for AzureCosmosStore {
+    async fn get(&self, key: &str) -> Result<Vec<u8>, Error> {
+        let pair = self.get_pair(key).await?;
+        Ok(pair.value)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
+        let pair = Pair {
+            id: key.to_string(),
+            value: value.to_vec(),
+        };
+        self.client
+            .lock()
+            .await
+            .create_document(pair)
+            .is_upsert(true)
+            .await
+            .map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Error> {
+        let pair = self.get_pair(key).await?;
+        let document_client = self
+            .client
+            .lock()
+            .await
+            .document_client(&pair.id, &pair.id)
+            .map_err(log_error)?;
+        document_client.delete_document().await.map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn exists(&self, key: &str) -> Result<bool, Error> {
+        Ok(self.get_keys().await?.contains(&key.to_string()))
```

Comment:
I wonder if calling `get_pair` here is cheaper.

Reply:
Updated.